sparse_strips/vello_common/src/glyph.rs

```diff
@@ -0,0 +1,190 @@
+    outline::{HintingInstance, HintingOptions, OutlinePen},
+};
+use vello_api::kurbo::{Affine, BezPath, Vec2};
+
+pub use vello_api::glyph::*;
+
+/// A glyph prepared for rendering.
+#[derive(Debug)]
+pub enum PreparedGlyph {
+    /// A glyph defined by its outline.
+    Outline(OutlineGlyph),
+    // TODO: Image and Colr variants.
+}
+
+/// A glyph defined by a path (its outline) and a local transform.
+#[derive(Debug)]
+pub struct OutlineGlyph {
+    /// The path of the glyph.
+    pub path: BezPath,
+    /// The local transform of the glyph.
+    pub local_transform: Affine,
+}
+
+/// Trait for types that can render glyphs.
+pub trait GlyphRenderer {
+    /// Fill glyphs with the current paint and fill rule.
+    fn fill_glyphs(&mut self, glyphs: impl Iterator<Item = PreparedGlyph>);
+
+    /// Stroke glyphs with the current paint and stroke settings.
+    fn stroke_glyphs(&mut self, glyphs: impl Iterator<Item = PreparedGlyph>);
+}
+
+/// A builder for configuring and drawing glyphs.
+#[derive(Debug)]
+pub struct GlyphRunBuilder<'a, T: GlyphRenderer + 'a> {
+    run: GlyphRun,
+    renderer: &'a mut T,
+}
+
+impl<'a, T: GlyphRenderer + 'a> GlyphRunBuilder<'a, T> {
+    /// Creates a new builder for drawing glyphs.
+    pub fn new(font: Font, renderer: &'a mut T) -> Self {
+        Self {
+            run: GlyphRun {
+                font,
+                font_size: 16.0,
+                glyph_transform: None,
+                hint: true,
+                normalized_coords: Vec::new(),
+            },
+            renderer,
+        }
+    }
+
+    /// Set the font size in pixels per em.
+    pub fn font_size(mut self, size: f32) -> Self {
+        self.run.font_size = size;
+        self
+    }
+
+    /// Set the per-glyph transform. Can be used to apply skew to simulate italic text.
+    pub fn glyph_transform(mut self, transform: Affine) -> Self {
+        self.run.glyph_transform = Some(transform);
+        self
+    }
+
+    /// Set whether font hinting is enabled.
+    pub fn hint(mut self, hint: bool) -> Self {
+        self.run.hint = hint;
+        self
+    }
+
+    /// Set normalized variation coordinates for variable fonts.
+    pub fn normalized_coords(mut self, coords: Vec<NormalizedCoord>) -> Self {
+        self.run.normalized_coords = coords;
+        self
+    }
+
+    /// Consumes the builder and fills the glyphs with the current configuration.
+    pub fn fill_glyphs(self, glyphs: impl Iterator<Item = &'a Glyph>) {
+        self.renderer
+            .fill_glyphs(Self::prepare_glyphs(&self.run, glyphs));
+    }
+
+    /// Consumes the builder and strokes the glyphs with the current configuration.
+    pub fn stroke_glyphs(self, glyphs: impl Iterator<Item = &'a Glyph>) {
+        self.renderer
+            .stroke_glyphs(Self::prepare_glyphs(&self.run, glyphs));
+    }
+
+    fn prepare_glyphs(
+        run: &GlyphRun,
+        glyphs: impl Iterator<Item = &'a Glyph>,
+    ) -> impl Iterator<Item = PreparedGlyph> {
```

Comment:
This allocates a `BezPath` for each `PreparedGlyph`. As we don't look at multiple glyphs at the same time, we could instead reuse a single underlying `BezPath` allocation. Perhaps by using something like a [lending iterator](https://blog.rust-lang.org/2022/10/28/gats-stabilization.html#what-are-gats), or shuffling code around to prevent needing an iterator.

Reply:
Prevented per glyph path allocation in https://github.com/linebender/vello/pull/883/commits/c458e8c694a69de11c5ac8343f5ad5805223df90